### ISSUE TITLE
[🐛] When select any image crash app

### ISSUE THREAD

**Issue by nericode** [react-native-image-picker/react-native-image-picker#1648]
### How to repeat issue and example

1.- Lauch method launchImageLibrary
2.- Select any image
3.- Confirm & crash

### Solution

What needs to be done to address this issue?
No crash app when select any image.

### Additional Information

```
FATAL EXCEPTION: main
Process: com.copymaster, PID: 7518
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=13002, result=-1, data=Intent { flg=0x1 hwFlg=0x10 (has extras) }} to activity {com.copymaster/com.copymaster.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.net.Uri.getLastPathSegment()' on a null object reference
	at android.app.ActivityThread.deliverResults(ActivityThread.java:5471)
	at android.app.ActivityThread.handleSendResult(ActivityThread.java:5512)
	at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:51)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:149)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:103)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2386)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	at android.os.Looper.loop(Looper.java:213)
	at android.app.ActivityThread.main(ActivityThread.java:8178)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.net.Uri.getLastPathSegment()' on a null object reference
	at com.imagepicker.Utils.getResponseMap(Utils.java:330)
	at com.imagepicker.ImagePickerModule.onImageObtained(ImagePickerModule.java:145)
	at com.imagepicker.ImagePickerModule.onActivityResult(ImagePickerModule.java:176)
	at com.facebook.react.bridge.ReactContext.onActivityResult(ReactContext.java:308)
	at com.facebook.react.ReactInstanceManager.onActivityResult(ReactInstanceManager.java:758)
	at com.facebook.react.ReactDelegate.onActivityResult(ReactDelegate.java:90)
	at com.facebook.react.ReactActivityDelegate.onActivityResult(ReactActivityDelegate.java:112)
	at com.facebook.react.ReactActivity.onActivityResult(ReactActivity.java:68)
	at android.app.Activity.dispatchActivityResult(Activity.java:8413)
	at android.app.ActivityThread.deliverResults(ActivityThread.java:5464)
	... 11 more
```

Environment: 
"react": "16.13.1",
"react-native": "0.63.3",
"react-native-image-picker": "^3.3.2",

Phone Test: 
* Huawei

**Comment by ravirajn22** [comment]
Thanks for the report, can you check if it happens in other android device models? It will be helpful.

**Comment by ansh** [comment]
This happens for me too. Here is the Stack Trace from Sentry:

```
java.lang.RuntimeException: Illegal callback invocation from native module. This callback type only permits a single invocation from native code.
    at com.facebook.react.bridge.CallbackImpl.invoke(CallbackImpl.java:26)
    at com.imagepicker.ImagePickerModule.onActivityResult(ImagePickerModule.java:153)
    at com.facebook.react.bridge.ReactContext.onActivityResult(ReactContext.java:308)
    at com.facebook.react.ReactInstanceManager.onActivityResult(ReactInstanceManager.java:758)
    at com.facebook.react.ReactDelegate.onActivityResult(ReactDelegate.java:90)
    at com.facebook.react.ReactActivityDelegate.onActivityResult(ReactActivityDelegate.java:112)
    at com.facebook.react.ReactActivity.onActivityResult(ReactActivity.java:68)
    at android.app.Activity.dispatchActivityResult(Activity.java:7797)
    at android.app.ActivityThread.deliverResults(ActivityThread.java:5071)
    at android.app.ActivityThread.handleSendResult(ActivityThread.java:5120)
    at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49)
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199)
    at android.os.Handler.dispatchMessage(Handler.java:112)
    at android.os.Looper.loop(Looper.java:216)
    at android.app.ActivityThread.main(ActivityThread.java:7625)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)

java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=13002, result=0, data=null} to activity {app.myapp/app.myapp.MainActivity}: java.lang.RuntimeException: Illegal callback invocation from native module. This callback type only permits a single invocation from native code.
    at android.app.ActivityThread.deliverResults(ActivityThread.java:5078)
    at android.app.ActivityThread.handleSendResult(ActivityThread.java:5120)
    at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49)
    at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
    at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199)
    at android.os.Handler.dispatchMessage(Handler.java:112)
    at android.os.Looper.loop(Looper.java:216)
    at android.app.ActivityThread.main(ActivityThread.java:7625)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)
```

### Here is the breadcrumbs...
<img width="1200" alt="Screen Shot 2021-03-23 at 2 51 11 AM" src="https://user-images.githubusercontent.com/8207733/112060012-a4185700-8b82-11eb-8aa1-76a55613ae3e.png">


### Here is the distribution of the devices and other details. Of course, this issue only occurs on Android devices of every model. You can see that we have over 55 unique devices which have faced this issue @ravirajn22 
<img width="320" alt="Screen Shot 2021-03-23 at 2 51 55 AM" src="https://user-images.githubusercontent.com/8207733/112060098-bc887180-8b82-11eb-841f-609aa417b187.png">

**Comment by apoorv-kansal-trysolvio** [comment]
Similar issue is occurring with Redmi phones also.

**Comment by chgsilva** [comment]
same here

**Comment by Prat1143** [comment]
Facing the same issue - as soon as I select an image to upload the app crashes without giving any errors, but it's displaying the response in the console. I also tried using react-native-image-crop-picker, but it too made the app crash on image upload. I was first running the app using the command "npx react-native run-android". And when the app crashed it was only showing the response of the uploaded image and nothing else or info related to app crash. But then I switched to Android Studio and built the app from there and as soon as I uploaded the image the following error was displayed in the logcat-

{ height: 200,
width: 92,
type: 'image/jpeg',
fileName: 'rn_image_picker_lib_temp_54c4dc23-202f-45a0-94da-1b05f313f8ed.jpg',
fileSize: 13008,
uri: 'file:///data/user/0/com.abc/cache/rn_image_picker_lib_temp_54c4dc23-202f-45a0-94da-1b05f313f8ed.jpg' }
2021-03-26 19:44:46.889 7469-7469/com.abc E/AndroidRuntime: FATAL EXCEPTION: main
Process: com.abc, PID: 7469
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/168402 flg=0x1 (has extras) }} to activity {com.abc/com.abc.MainActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'void com.facebook.react.bridge.Callback.invoke(java.lang.Object[])' on a null object reference
at android.app.ActivityThread.deliverResults(ActivityThread.java:5078)
at android.app.ActivityThread.handleSendResult(ActivityThread.java:5120)
at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199)
at android.os.Handler.dispatchMessage(Handler.java:112)
at android.os.Looper.loop(Looper.java:216)
at android.app.ActivityThread.main(ActivityThread.java:7625)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void com.facebook.react.bridge.Callback.invoke(java.lang.Object[])' on a null object reference
at com.reactlibrary.RNPgReactNativeSdkModule$1.onActivityResult(RNPgReactNativeSdkModule.java:157)
at com.facebook.react.bridge.ReactContext.onActivityResult(ReactContext.java:308)
at com.facebook.react.ReactInstanceManager.onActivityResult(ReactInstanceManager.java:758)
at com.facebook.react.ReactDelegate.onActivityResult(ReactDelegate.java:90)
at com.facebook.react.ReactActivityDelegate.onActivityResult(ReactActivityDelegate.java:112)
at com.facebook.react.ReactActivity.onActivityResult(ReactActivity.java:68)
at android.app.Activity.dispatchActivityResult(Activity.java:7797)
at android.app.ActivityThread.deliverResults(ActivityThread.java:5071)
at android.app.ActivityThread.handleSendResult(ActivityThread.java:5120) 
at android.app.servertransaction.ActivityResultItem.execute(ActivityResultItem.java:49) 
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199) 
at android.os.Handler.dispatchMessage(Handler.java:112) 
at android.os.Looper.loop(Looper.java:216) 
at android.app.ActivityThread.main(ActivityThread.java:7625) 
at java.lang.reflect.Method.invoke(Native Method) 
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524) 
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987) 
2021-03-26 19:44:46.892 7469-7469/com.abc D/e: CrashLogs : {"ts":"2021-03-26 19:44:46","app_id":"UNKNOWN","package":"UNKNOWN","android_id":"UNKNOWN","platform":"Linux","os_version":"28","device_model":"COR-AL00","device_name":"HONOR","sdk_version":"1.6.0","network_type":"UNKNOWN","stack":"com.reactlibrary.RNPgReactNativeSdkModule$1.onActivityResult(RNPgReactNativeSdkModule.java:157)","cause":"java.lang.NullPointerException: Attempt to invoke interface method 'void com.facebook.react.bridge.Callback.invoke(java.lang.Object[])' on a null object reference"}

Am I missing something or doing something wrong???

**Comment by alexandruenache19** [comment]
Any solution?

**Comment by codal-mpawar** [comment]
Facing same issue for android.
1. When selecting image from library getting crash and unknown urls and not getting mp4 kind of url.
2. When selecting video from library getting crash.

**Comment by Maddumage** [comment]
Same issue in Samsung M20 and SM-A530F devices. iOS working perfectly.

**Comment by MinWangWTFSSD** [comment]
i have the same issue on SM C5010 android 8.0.0 and resolve by:
1. create empty custom FileProvider like this 
 ```java
import androidx.core.content.FileProvider;

public class MyFileProvider extends FileProvider {
}
```
2. in `AndroidManifest.xml` add a `provider` tag like this 
```xml
<provider 
    android:authorities="${applicationId}.provider"
    android:name=".MyFileProvider"
    android:exported="false"
    tools:replace="android:authorities"
    android:grantUriPermissions="true" >
      <meta-data
          android:name="android.support.FILE_PROVIDER_PATHS"
          android:resource="@xml/file_paths"
          tools:replace="android:resource"/>
</provider>
```

**Comment by joaojeefer** [comment]
The same problem is happening to me.
Could anyone solve it?

**Comment by Dem0n13** [comment]
The same issue:
```
Fatal Exception: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/298663 flg=0x1 }} to activity {com.starbyface/com.starbyface.MainActivity}: java.lang.RuntimeException: Illegal callback invocation from native module. This callback type only permits a single invocation from native code.
       at android.app.ActivityThread.deliverResults(ActivityThread.java:4192)
       at android.app.ActivityThread.handleSendResult(ActivityThread.java:4235)
       at android.app.ActivityThread.-wrap20(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1584)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6317)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:872)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:762)

Caused by java.lang.RuntimeException: Illegal callback invocation from native module. This callback type only permits a single invocation from native code.
       at com.facebook.react.bridge.CallbackImpl.invoke(CallbackImpl.java:26)
       at com.imagepicker.ImagePickerModule.onImageObtained(ImagePickerModule.java:145)
       at com.imagepicker.ImagePickerModule.onActivityResult(ImagePickerModule.java:176)
       at com.facebook.react.bridge.ReactContext.onActivityResult(ReactContext.java:308)
       at com.facebook.react.ReactInstanceManager.onActivityResult(ReactInstanceManager.java:758)
       at com.facebook.react.ReactDelegate.onActivityResult(ReactDelegate.java:90)
       at com.facebook.react.ReactActivityDelegate.onActivityResult(ReactActivityDelegate.java:112)
       at com.facebook.react.ReactActivity.onActivityResult(ReactActivity.java:68)
       at android.app.Activity.dispatchActivityResult(Activity.java:7010)
       at android.app.ActivityThread.deliverResults(ActivityThread.java:4188)
       at android.app.ActivityThread.handleSendResult(ActivityThread.java:4235)
       at android.app.ActivityThread.-wrap20(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1584)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6317)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:872)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:762)
```

As I investigated: any android versions (7-10), any vendors. I comes from the next call:
```
ImagePicker.launchImageLibrary({ mediaType: 'photo', includeBase64: false }, response => {})
```
My idea is the cause of problem is null returning from getAppSpecificStorageUri. Here are some examples of ResultInfo's from Crash reports:
```
{who=null, request=13001, result=-1, data=null}
{who=null, request=13002, result=-1, data=Intent { dat=content://com.miui.gallery.open/raw//storage/emulated/0/Snapchat/Snapchat-164691440.jpg typ=image/jpeg flg=0x1 }}
{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/298663 flg=0x1 }}
{who=null, request=13002, result=0, data=null}
{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/2058 flg=0x1 (has extras) }}
{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/1441 flg=0x1 (has extras) }}
{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/2571 flg=0x1 (has extras) }}
{who=null, request=13002, result=-1, data=Intent { dat=content://media/external/images/media/2038 flg=0x1 (has extras) }}
```
It's strange to see this reports from "result=-1".
Note that I've already throttle button clicks by 1 second (I tried to prevent button multiclicking). Hope that it will be useful.

**Comment by Dem0n13** [comment]
*MinWangWTFSSD Could you provide your @xml/file_paths ? Does it really fixed this problem?
As I see react-native0-image-picker already provides it:

AndroidManifest.xml
```
<?xml version="1.0" encoding="utf-8"?>
<manifest
  xmlns:android="http://schemas.android.com/apk/res/android"
  package="com.imagepicker"
  >
    <application>
      <provider
            android:name=".ImagePickerProvider"
            android:authorities="${applicationId}.imagepickerprovider"
            android:exported="false"
            android:grantUriPermissions="true">
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/imagepicker_provider_paths" />
        </provider>
    </application>
</manifest>
```
imagepicker_provider_paths.xml
```
<?xml version="1.0" encoding="utf-8"?>
<paths>
  <cache-path name="cacheDir" path="." />
</paths>
```

**Comment by akhilumapathe** [comment]
Issue still unresolved.

**Comment by MinWangWTFSSD** [comment]
@Dem0n13  this is my file_paths.xml:
```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <paths>
        <external-path name="apks" path="."/>


    </paths>
</resources>
```

**Comment by Maddumage** [comment]
@MinWangWTFSSD is that worked for you?

**Comment by MinWangWTFSSD** [comment]
@Maddumage  @Dem0n13  try to recheck `WRITE_EXTERNAL_STORAGE`  permission? in my project  is use `"react-native-permissions": "^2.2.2"`  to check `WRITE_EXTERNAL_STORAGE`  permission. follow is my select image code:
```typescript
import {Alert,Platform} from 'react-native'
import ImagePicker from "react-native-image-picker";
import { PERMISSIONS, RESULTS, request } from 'react-native-permissions';

export const showImagePicker = (title:string): Promise<{
    customButton: string;
    didCancel: boolean;
    error: string;
    data: string;
    uri: string;
    origURL?: string;
    isVertical: boolean;
    width: number;
    height: number;
    fileSize: number;
    type?: string;
    fileName?: string;
    path?: string;
    latitude?: number;
    longitude?: number;
    timestamp?: string;
}> => {
    const options = {
        ...you options...

        cameraType: 'back',
        mediaType: 'photo',
        videoQuality: 'high',
        angle: 0,
        allowsEditing: false,
        noData: false,
        saveToPhotos: true,
        includeBase64: false,
    };
    return new Promise(async (resolve, reject) => {

        const requestPermission = (camera: boolean) => {
            return request(
                Platform.select({
                    android: camera ? PERMISSIONS.ANDROID.CAMERA : PERMISSIONS.ANDROID.READ_EXTERNAL_STORAGE,
                    ios: camera ? PERMISSIONS.IOS.CAMERA : PERMISSIONS.IOS.PHOTO_LIBRARY,
                })).then(res => {
                    return res == RESULTS.GRANTED
                })
        }

        const requestWritePermission = () => {
            return request(
                Platform.select({
                    android: PERMISSIONS.ANDROID.WRITE_EXTERNAL_STORAGE,
                    ios: PERMISSIONS.IOS.MEDIA_LIBRARY
                })).then(res => {
                    return res == RESULTS.GRANTED
                })
        }



        Alert.alert(title, '', [
            {
                text: 'cancel',
                style: 'cancel',
                onPress: () => {
                    reject(new Error('user cancled'))
                }
            },
            {
                text: "Library",
                onPress: async () => {

                    const per = await requestPermission(false)
                    const per1 = await requestWritePermission()
                    if (!per || !per1) return resolve(null)
                    ImagePicker.launchImageLibrary(options, (response: any) => {
                        if (!response || response.errorCode) {
                            return resolve(null)
                        }
                        resolve(response)
                    })
                }
            },
            {
                text: "Camera",
                onPress: async () => {
                    const per = await requestPermission(true)
                    const per1 = await requestWritePermission()
                    if (!per || !per1) return resolve(null)
                    ImagePicker.launchCamera(options, (response: any) => {
                        if (!response || response.errorCode) {
                            return resolve(null)
                        }
                        resolve(response)
                    })
                }
            }
        ])
    });

};
```
Hope useful

**Comment by ludovic-coder** [comment]
It may be obvious but be sure your button handle double taps. It's ease to reproduce this issue double tapping the button. With android, native modules can be mounted many times, it's not the case with iOS. Maybe your customers tap the button, the native modal take to many time so show and clicked again. Or maybe your UI is not good enough and the "is tapping" state is not very clear for the user.

**Comment by systemride** [comment]
> It may be obvious but be sure your button handle double taps. It's ease to reproduce this issue double tapping the button. With android, native modules can be mounted many times, it's not the case with iOS. Maybe your customers tap the button, the native modal take to many time to show and clicked again. Or maybe your UI is not good enough and the "is tapping" state is not very clear for the user.

The double-tapping is onto something. I am able to reproduce this in the native image library:

- Open the image picker.
- Navigate into one of your photo folders.
- Take both thumbs and tap multiple times on two different images very quickly.
- App will crash with this error message.

This is on a Google Pixel 3A device.

**Comment by Dem0n13** [comment]
@systemride, Thank you for suggestion, do you know how to fix this problem? Any workarounds?

**Comment by ozberkctn** [comment]
Same issue

**Comment by JosephVasse** [comment]
Having the same problem here. 
App crashes when I select the picture on Android.

Any workaround ?

**Comment by systemride** [comment]
Per my comment above, it was actually double-tapping the button to open the image library that was causing the issue (@ ludovicj44 was right on).

There are many ways to resolve this but I did so by adding a property on the button's local state: `isDisabledAfterFirstPress` and then after the first press it sets this to true. In my case, the button unmounts after press as well which is important otherwise the button could get into a stuck state.

Hope this helps.

**Comment by yashSrathore** [comment]
Same problem app crashing after clicking image im pressing button one time only.

**Comment by jacob-san** [comment]
```
  const pickImage = () => async () => {
    const options = {
      mediaType: 'photo',
      cameraType: 'back',
    };
    if(pickerInitiated){
      // User initiated picker already, Prevent multiple invocation untill dismissed
      return;
    }
    setPickerInitiated(true);
    const grantedPerms = await grantedPermission(permissionType);
    if (grantedPerms) {
      try {
        launchImageGallery(options, async res => {
          setPickerInitiated(false);
          if (res.didCancel || res.errorCode) {
            return;
          }
          const response = res?.assets[0];
          let selectedFile = {
            fileSize: response.fileSize,
            uri: response.uri,
            type: response.type,
            name: response.name,
          };
          setSelectedFile(selectedFile);
        });
      } catch (error) {
        setPickerInitiated(false);
      }
    }
  };
```
Hope this helps someone to prevent multiple invocation.

**Comment by aaftabali1** [comment]
I think you need to close the model and after that you are able to click the image from camera.

**Comment by Desintegrator** [comment]
You can fix ios "illegal callback invocation from navite module" as i suggest [here](https://github.com/react-native-image-picker/react-native-image-picker/pull/1922)

**Comment by yorickshan** [comment]
same issue

**Comment by lukebrandonfarrell** [comment]
Fixed in #1948
